client/src/commands/run.ts

```diff
@@ -54,26 +59,33 @@ function getCode(selected = false, uri?: Uri): string {
     }
   }
   let code = "";
+  let selections: ReadonlyArray<Selection>;
   if (selected) {
     // run selected code if there is one or more non-empty selections, otherwise run all code
 
+    selections = editor.selections;
     // since you can have multiple selections, append the text for each selection in order of selection
     // note: selection ranges can be empty (ex. just a carat)
     for (const selection of editor.selections) {
       const selectedText: string = doc.getText(selection);
-      code += selectedText;
+      code += selectedText + "\n";
     }
-    // if no non-whitespace characters are selected, treat as no selection and run all code
-    if (code.trim().length === 0) {
-      code = doc?.getText();
-    }
-  } else {
+  }
+
+  if (code.trim().length === 0) {
     code = doc?.getText();
+    selections = [new Selection(0, 0, doc.lineCount - 1, 0)];
   }
+
+  code = addLocators(code);
```

Comment:
add/remove look pretty separated, it's not clear if a code is the real code at a given place.
I would suggest to return an offset from wrapping.

Reply:
The problem is that it is unknow how many extra code will be added into the code from source file, so adding a locator to mark the beginning of code from source file, then it is easy to get the offset by that locator.